<a href="https://colab.research.google.com/github/Abdullah2020/Abdullah2020/blob/master/ML_lab2_Abdullahi_Isa.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Part II : Practical application graded !**

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

print("Setup Completed...!!!")

Setup Completed...!!!


In [5]:
auto = pd.read_csv("/content/auto.txt", sep=";")

In [6]:
auto.head()

mpg  cylinders  displacement  horsepower  weight  acceleration  year  \
0  18.0          8         307.0         130    3504          12.0    70   
1  15.0          8         350.0         165    3693          11.5    70   
2  18.0          8         318.0         150    3436          11.0    70   
3  16.0          8         304.0         150    3433          12.0    70   
4  17.0          8         302.0         140    3449          10.5    70   

   origin                       name  
0       1  chevrolet chevelle malibu  
1       1          buick skylark 320  
2       1         plymouth satellite  
3       1              amc rebel sst  
4       1                ford torino

In [14]:
auto.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 10 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   mpg           300 non-null    float64
 1   cylinders     300 non-null    int64  
 2   displacement  300 non-null    float64
 3   horsepower    300 non-null    int64  
 4   weight        300 non-null    int64  
 5   acceleration  300 non-null    float64
 6   year          300 non-null    int64  
 7   origin        300 non-null    int64  
 8   name          300 non-null    object 
 9   mpg01         300 non-null    uint8  
dtypes: float64(3), int64(5), object(1), uint8(1)
memory usage: 21.5+ KB


**Comment:**

As shown from the above table, there are 10 columns in the auto dataset (index inclusive). One of the variable is categorical (labelled as 'object') while the remaining 9 are numerical (labelled as 'int64' or 'float64').

**2. Graded Question:**

Create a binary variable, denoted mpg01, that is worth 1 if mpg contains a value above its
median, and a 0 if mpg contains a value below its median. You can use the method median()
and the function get_dummies(). Then, create a single data set containing both mpg01 and the
other Auto variables.

In [7]:
mpg_median = auto["mpg"].median()
mpg01 = pd.get_dummies(auto["mpg"] > mpg_median, drop_first=True)

auto["mpg01"] = mpg01

In [8]:
auto.describe()

mpg   cylinders  displacement  horsepower       weight  \
count  300.000000  300.000000    300.000000  300.000000   300.000000   
mean    23.699333    5.433333    192.338333  104.050000  2945.846667   
std      7.963814    1.717183    105.453661   38.959127   849.694403   
min      9.000000    3.000000     68.000000   46.000000  1613.000000   
25%     17.000000    4.000000     98.000000   75.000000  2212.500000   
50%     23.600000    4.000000    140.000000   91.500000  2730.000000   
75%     29.925000    8.000000    275.750000  129.000000  3526.250000   
max     46.600000    8.000000    455.000000  230.000000  4997.000000   

       acceleration       year      origin       mpg01  
count    300.000000  300.00000  300.000000  300.000000  
mean      15.460000   75.93000    1.616667    0.500000  
std        2.729389    3.67015    0.827853    0.500835  
min        8.000000   70.00000    1.000000    0.000000  
25%       13.500000   73.00000    1.000000    0.000000  
50%       15.400000   76.00000    1.000000    0.500000  
75%       17.000000   79.00000    2.000000    1.000000  
max       23.500000   82.00000    3.000000    1.000000

**Comment:**

The above output gives an important summary of the auto dataset statistics of the numerical variables such as the mean, median (i.e. 50%), minimum, maximum (measures of the central tendency), and standard deviation values. It also shows the Interquartile Range (IQR) using the 25th, and 75th percentile values. 

Therefore, from the output, we can infer that the average ‘miles per gallon’ is 23.7, the average ‘year’ is about 75.9 years, while the displacement average is 192.3 etc. However, the describe() function only prints the statistical values of quantitative or numerical variables and ignore the categorical column(s). Though, both types of variables can we display using include=’all’ argument, which can be added inside describe() function.


In [13]:
print("The median of mpg is :", auto["mpg"].median())

The median of mpg is : 23.6


**Comment:**

The Mean of mpg represents the arithmetic average of the data. The line of code above prints the mean of the numerical variables in the mpg data. From the output, we can infer that the average miles per gallon of the fuel consumption of the car is 23.6,

**3. Graded Question:**

You will fit the classifiers studied in the lecture using the train data and evaluate the quality
of your models using the test data. That is why the whole data set was split into two sub-
sets. To this end, you will find in moodle 4 files x_train.csv, x_test.csv, y_train.csv and
y_test.csv. 

Import the four datasets.

In [15]:
x_test = pd.read_csv("/content/x_test.csv", index_col = 0)
x_train = pd.read_csv("/content/x_train.csv", index_col = 0)
y_test = pd.read_csv("/content/y_test.csv", index_col = 0)
y_train = pd.read_csv("/content/y_train.csv", index_col = 0)

**Graded Question:**

How many observations are in each dataset?

In [19]:
print("There are {} obervations in x_test dataset" .format(x_test.shape[0]))
print("There are {} obervations in x_train dataset" .format(x_train.shape[0]))
print("There are {} obervations in y_test dataset" .format(y_test.shape[0]))
print("There are {} obervations in y_train dataset" .format(y_train.shape[0]))


There are 60 obervations in x_test dataset
There are 240 obervations in x_train dataset
There are 60 obervations in y_test dataset
There are 240 obervations in y_train dataset


**Comment:**

As seen above, x_test, x_train, y_test, y_train have 60, 240, 60 and 240 observations respectively.

# **2.2 Logistic regression**

# **2.3 K-Nearest Neighbors**

You can build a K-Nearest Neighbors model using the ***KNeighborsClassifier()*** function, which is part of the neighbors submodule of (sklearn). For instance, to train KNN with K = 1 classifier in training sets *y_train* and *x_train* and then make predictions with the test sets *y_test* and *x_test* run the following commands :

**Import KNN library from sklearn**

In [20]:
from sklearn.neighbors import KNeighborsClassifier

**KNN Classification model when k=1**

In [21]:
#define model when k=1
neigh = KNeighborsClassifier(n_neighbors=1)

#fit model
neigh.fit(x_train, np.ravel(y_train))

#make prediction on x_test dataframe
yhat_knn = neigh.predict(x_test)

In [25]:
#Import scikit-learn metrics module for accuracy calculation
from sklearn import metrics

# Model Accuracy, how often is the classifier correct?
print("Accuracy:",metrics.accuracy_score(y_test, yhat_knn))

Accuracy: 0.8666666666666667


In [37]:
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score

result = confusion_matrix(y_test, yhat_knn)
print("Confusion Matrix:")

print(result)
result1 = classification_report(y_test, yhat_knn)

print("Classification Report:",)
print (result1)
result2 = accuracy_score(y_test,yhat_knn)
print("Accuracy:", round(result2,4))

Confusion Matrix:
[[29  5]
 [ 3 23]]
Classification Report:
              precision    recall  f1-score   support

           0       0.91      0.85      0.88        34
           1       0.82      0.88      0.85        26

    accuracy                           0.87        60
   macro avg       0.86      0.87      0.87        60
weighted avg       0.87      0.87      0.87        60

Accuracy: 0.8667


**Graded Question:**

Fit a KNN classifier for the following values of K in th element of {1; 3; 5; 17; 51; 75; 101} and calculate the test error. 

What value of K would you choose? For this value of K calculate the performance indicators you calculated in the previous section.

**k_values = [1, 3, 5, 17, 51, 75, 101]**

**KNN Classification model when k=3**

In [31]:
# model with k=3
knn_3 = KNeighborsClassifier(n_neighbors=3)

# fit the model to the training set
knn_3.fit(x_train, np.ravel(y_train))

# predict on the test-set
y_pred_3 = knn_3.predict(x_test)

print('Model accuracy score with k=3 : {0:0.4f}'. format(accuracy_score(y_test, y_pred_3)))

Model accuracy score with k=3 : 0.8833


**KNN Classification model when k=5**

In [32]:
# model with k=5
knn_5 = KNeighborsClassifier(n_neighbors=5)

# fit the model to the training set
knn_5.fit(x_train, np.ravel(y_train))

# predict on the test-set
y_pred_5 = knn_5.predict(x_test)

print('Model accuracy score with k=5 : {0:0.4f}'. format(accuracy_score(y_test, y_pred_5)))

Model accuracy score with k=5 : 0.8333


**KNN Classification model when k=17**

In [33]:
# model with k=17
knn_17 = KNeighborsClassifier(n_neighbors=17)

# fit the model to the training set
knn_17.fit(x_train, np.ravel(y_train))

# predict on the test-set
y_pred_17 = knn_17.predict(x_test)

print('Model accuracy score with k=17 : {0:0.4f}'. format(accuracy_score(y_test, y_pred_17)))

Model accuracy score with k=17 : 0.8500


**KNN Classification model when k=51**

In [34]:
# model with  k=51
knn_51 = KNeighborsClassifier(n_neighbors=51)

# fit the model to the training set
knn_51.fit(x_train, np.ravel(y_train))

# predict on the test-set
y_pred_51 = knn_51.predict(x_test)

print('Model accuracy score with k=51 : {0:0.4f}'. format(accuracy_score(y_test, y_pred_51)))

Model accuracy score with k=51 : 0.8500


**KNN Classification model when k=75**

In [35]:
# model with  k=75
knn_75 = KNeighborsClassifier(n_neighbors=75)

# fit the model to the training set
knn_75.fit(x_train, np.ravel(y_train))

# predict on the test-set
y_pred_75 = knn_75.predict(x_test)

print('Model accuracy score with k=75 : {0:0.4f}'. format(accuracy_score(y_test, y_pred_75)))

Model accuracy score with k=75 : 0.8333


**KNN Classification model when k=101**

In [36]:
# model with  k=101
knn_101 = KNeighborsClassifier(n_neighbors=101)

# fit the model to the training set
knn_101.fit(x_train, np.ravel(y_train))

# predict on the test-set
y_pred_101 = knn_101.predict(x_test)

print('Model accuracy score with k=101 : {0:0.4f}'. format(accuracy_score(y_test, y_pred_101)))

Model accuracy score with k=101 : 0.8333


**Comment:**

As seen from the above k-NN values, the model accuracy score with **k=1** is **0.8667** while when **k = 3, 5, 17, 51, 75 and 101**, it returns an accuracy scores of **0.8667, 0.8833, 0.8333, 0.8500, 0.8500, 0.8333, and 0.8333** respectively.

Thus, based on the above analysis, we can conclude that the model accuracy is good with k = 3 being the best performing model (88.33%). Similarly, the boundary is highly linear and the best values for k should be relatively SMALL to avoid overfitting.

# **2.4 Discriminant Analysis**

The *LinearDiscriminantAnalysis()* and the *QuadraticDiscriminantAnalysis()* functions from the class sklearn.discriminant_analysis allow to perform Linear and Quadratic discriminant Analysis respectively. For instance, to perform LDA on the train set and predict the target variable on the test set you can run the following code.

**Import Discriminant Analysis library from sklearn**

In [38]:
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis

In [39]:
#define model
lda = LinearDiscriminantAnalysis()

#fit model
model_lda = lda.fit(x_train, np.ravel(y_train))

#make prediction
yhat_lda = model_lda.predict(x_test)

In [40]:
print(model_lda.priors_)

[0.48333333 0.51666667]


In [41]:
print(model_lda.means_)

[[   6.74137931  270.73275862  129.24137931 3585.03448276]
 [   4.15322581  112.25403226   77.85483871 2302.2016129 ]]


**Graded Question:**

1 - Interpret the prior probabilities as well as the group means.

**Comment:**

As seen above, the prior probabilities of group represent the proportions of each features in the training set. Therefore, 51.7% of all observations in the training set were ........


The group means help to display the mean values for each predictor variable for each car.

**Graded Question:**

2- What is the test error of the model obtained with the LDA classifier ?